modules/ding_innskraning/ding_innskraning.module

```diff
@@ -0,0 +1,780 @@
+  }
+
+  // Check destination.
+  $destination = trim($xmlDoc->documentElement->getAttribute('Destination'));
+  $currentURL = $base_url . '/' . request_path();
+  if ($destination !== $currentURL) {
+    throw new Exception("The response was received at $currentURL instead of $destination");
+  }
+
+  $user_agent = _ding_innskraning_get_user_agent($xmlDoc);
+  if ($user_agent !== $_SERVER['HTTP_USER_AGENT']) {
+    throw new Exception('Request was done from another browser');
+  }
+
+  // @TODO: why is this commented?
+  // check the issuer
+  /*$issuer = $xmlDoc->getElementsByTagName('Issuer');
+
+    if ($issuer->length < 1) {
+    throw new Exception('There is no issuer info in the document');
+    } elseif ($issuer->item(0)->textContent != 'Innskraning') {
+    $current_issuer = $issuer->item(0)->textContent;
+    throw new Exception("The issuer is $current_issuer instead of Innskraning");
+    }*/
+
+  // Check the digital signature.
+  $objXMLSecDSig = new XMLSecurityDSig();
+  $objDSig = $objXMLSecDSig->locateSignature($xmlDoc);
+
+  if ($objDSig == NULL) {
+    throw new Exception("Cannot locate Signature Node");
+  }
+  $objXMLSecDSig->canonicalizeSignedInfo();
+  $objXMLSecDSig->idKeys = array('ID');
+  $objXMLSecDSig->idNS = array('wsu' => 'http://docs.oasisopen.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd');
+
+  $retVal = $objXMLSecDSig->validateReference();
+
+  if ($retVal == NULL) {
+    throw new Exception("Reference Validation Failed");
+  }
+
+  // Check token to ensure that the auth request originated from us. And
+  // island.is doesn't keep the casing, so fold to lower case.
+  $saml_token = drupal_strtolower(_ding_innskraning_get_attribute_by_name($xmlDoc, 'AuthID'));
+  $token = drupal_strtolower($_SESSION['ding_innskraning_token']);
+  if ($token && $saml_token !== $token) {
+    watchdog(
+      'ding_innskraning',
+      "Token validation failed, expected @token, got @saml_token.",
+      array('@token' => $token, '@saml_token' => $saml_token),
+      WATCHDOG_WARNING
+    );
+    throw new Exception("Integrity check failed.");
+  }
+
+  if (!_ding_innskraning_verify_date($xmlDoc)) {
+    throw new Exception("Conditions not valid.");
+  }
+
+  // @TODO: why is this commented?
+  /*if (!_ding_innskraning_verify_conditions($xmlDoc, _ding_innskraning_get_client_ip())){
+    throw new Exception("Invalid client ip.");
+    }*/
+  $objKey = $objXMLSecDSig->locateKey();
+
+  if (!$objKey) {
+    throw new Exception("Key not found");
+  }
+  $key = NULL;
+
+  $objKeyInfo = XMLSecEnc::staticLocateKeyInfo($objKey, $objDSig);
+
+  if (!_ding_innskraning_verify_cert($objKeyInfo)) {
+    throw new Exception("Certificate not valid");
+  }
+
+  if (!$objKeyInfo->key && empty($key)) {
+    $key = _ding_innskraning_get_key(TRUE);
+    $objKey->loadKey($key, TRUE);
+  }
+
+  if (!$objXMLSecDSig->verify($objKey)) {
+    throw new Exception("Signature invalid!");
+  }
+  else {
+    return "Signature valid.";
```

Comment:
This returns a `string`? That seems very odd.

Reply:
Fixed